vm/ignite/src/micro_code/assign.rs

```diff
@@ -14,12 +14,28 @@ use crate::{Runtime, VmError};
 /// # Errors
 ///
 /// If the stack is empty.
+use std::rc::Rc;
 pub fn assign(rt: &mut Runtime, sym: Symbol) -> Result<()> {
```

Comment:
Added this to handle nested assignment, e.g
```
let x = 2; 
let y = 0; 
{ 
    let x = 3; 
    y = 4; 
} 
x+y
```

Previously it would treat ```y = 4``` in the nested block as ```let y = 4``` and just make a new y variable, but the expected behaviour is to assign in the nearest scope that has the variable uninit/with a value, or failing that the global scope where parent is none. 

So ```x+y``` evals to 6 since ```let x = 3;``` is variable shadowing but ```y = 4``` assigns to y in the outer scope.

Reply:
Realised I can fix the issue with the variable not being declared anywhere with this at the end of assign, plus compiling the entire program with enter/exit scope:
```
if env_ptr.borrow().parent.is_none() && !env_ptr.borrow().env.contains_key(&sym){
        return Err(VmError::SymbolNotFound(sym.to_string()).into());
    }
```

i.e these would throw symbol not found at runtime:
```
x = 10; // because x was not declared anywhere
```

```
let y =10; { x = 10; } // because x was not declared anywhere
```

Otherwise without the check those would just assign in the global scope. 

But the issue is with this implementation it fails these cases in micro_code/runtime:
```
micro_code::assign::tests::test_assign
micro_code::assign::tests::test_assign_with_parent
runtime::tests::test_assignment
```

Should we change the cases or allow this global assignment without let? Or do you have any other ideas 🤔 I'm not sure

Reply:
This is a bug in the `assign` microcode and it is fixed in https://github.com/crabscript/rustscript/pull/37

Reply:
Thanks a lot!